File: 0077-combinations/javascript/Solution.js
/**
 * @param {number} n
 * @param {number} k
 * @return {number[][]}
 */
var combine = function(n, k) {
    let result = []
    let path = []
    let backtrack = step => {
        if (path.length == k) {
            result.push([...path])
            return
        }
        if (step == n+1) {
            return
        }
        backtrack(step+1)
        path.push(step)
        backtrack(step+1)
        path.pop()
    }
    backtrack(1)
    return result
};

File: 0077-combinations/python/Solution.py
class Solution:
    def __init__(self):
        self.result = []

    def combine(self, n: int, k: int) -> List[List[int]]:
        self.backtrack(n, k, 1, [])
        return self.result

    def backtrack(self, n, k, step, path):
        if len(path) == k:
            self.result.append(path)
            return
        if step == n + 1:
            return
        self.backtrack(n, k, step+1, path)
        path.append(step)
        self.backtrack(n,k,step+1, path)
        path.pop()
